services/pipeline/src/test/java/tech/pegasys/pantheon/services/pipeline/ProcessingStageTest.java

```diff
@@ -68,22 +70,42 @@ public void shouldCallSingleStepStageForEachInput() {
 
   @Test
   public void shouldFinalizeSingleStepStageAndCloseOutputPipeWhenInputCloses() {
+    when(singleStep.attemptFinalization(outputPipe)).thenReturn(true);
     inputPipe.close();
 
     stage.run();
 
-    verify(singleStep).finalize(outputPipe);
+    verify(singleStep).attemptFinalization(outputPipe);
+    verifyNoMoreInteractions(singleStep);
+    assertThat(outputPipe.isOpen()).isFalse();
+  }
+
+  @Test
+  public void shouldAbortIfPipeIsCancelledWhileAttemptingToFinalise() {
+    when(singleStep.attemptFinalization(outputPipe))
+        .thenAnswer(
+            invocation -> {
+              inputPipe.abort();
+              return false;
+            });
+    inputPipe.close();
```

Comment:
Why do you need to close the inputPipe here?

Reply:
Because otherwise the processor won't enter the finalisation stage, it will wait forever for the next input.